detectors/resources-support/src/main/java/com/google/cloud/opentelemetry/detectors/GCPPlatformDetector.java

```diff
@@ -0,0 +1,115 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.opentelemetry.detectors;
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+public class GCPPlatformDetector {
+  public static final GCPPlatformDetector DEFAULT_INSTANCE = new GCPPlatformDetector();
+
+  private final GCPMetadataConfig metadataConfig;
+  private final EnvironmentVariables environmentVariables;
+
+  // for testing only
+  GCPPlatformDetector(GCPMetadataConfig metadataConfig, EnvironmentVariables environmentVariables) {
+    this.metadataConfig = metadataConfig;
+    this.environmentVariables = environmentVariables;
+  }
+
+  private GCPPlatformDetector() {
+    this.metadataConfig = GCPMetadataConfig.DEFAULT_INSTANCE;
+    this.environmentVariables = EnvironmentVariables.DEFAULT_INSTANCE;
+  }
+
+  // Detects the GCP platform on which the application is running
+  public DetectedPlatform detectPlatform() {
+    if (!isRunningOnGcp()) {
+      return generateDetectedPlatform(SupportedPlatform.UNKNOWN_PLATFORM);
+    }
+    Function<EnvironmentVariables, Optional<SupportedPlatform>> detectGKE =
+        environmentVariables ->
+            environmentVariables.get("KUBERNETES_SERVICE_HOST") != null
+                ? Optional.of(SupportedPlatform.GOOGLE_KUBERNETES_ENGINE)
+                : Optional.empty();
+    Function<EnvironmentVariables, Optional<SupportedPlatform>> detectGCR =
+        environmentVariables ->
+            environmentVariables.get("K_CONFIGURATION") != null
+                    && environmentVariables.get("FUNCTION_TARGET") == null
+                ? Optional.of(SupportedPlatform.GOOGLE_CLOUD_RUN)
+                : Optional.empty();
+    Function<EnvironmentVariables, Optional<SupportedPlatform>> detectGCF =
+        environmentVariables ->
+            environmentVariables.get("FUNCTION_TARGET") != null
+                ? Optional.of(SupportedPlatform.GOOGLE_CLOUD_FUNCTIONS)
+                : Optional.empty();
+    Function<EnvironmentVariables, Optional<SupportedPlatform>> detectGAE =
+        environmentVariables ->
+            environmentVariables.get("GAE_SERVICE") != null
+                ? Optional.of(SupportedPlatform.GOOGLE_APP_ENGINE)
+                : Optional.empty();
+
+    // Order of detection functions matters here
+    Stream<Function<EnvironmentVariables, Optional<SupportedPlatform>>> platforms =
+        Stream.of(detectGKE, detectGCR, detectGCF, detectGAE);
```

Comment:
Shouldn't GCF be before GCR?

Reply:
I did not look at the spec again, but I was going by the code we already had - in which [we checked for GCR before GCF](https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/blob/de1d4b4dbb25503afe271d28f68f20c81740d0b8/detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors/GCPResource.java#L65).

Is there a reason why GCF should be checked before GCR ?